From 6f8c86de254562b39b896a5bbbfcf51cef9cc7b7 Mon Sep 17 00:00:00 2001
From: Nitish Mehta <nitishm@vivi.io>
Date: Mon, 5 Oct 2026 16:06:01 +1100
Subject: [PATCH] gstsouphttpsrc: use soup_auth_is_for_proxy to pick proxy
 credentials

With libsoup 3, a 407 on a CONNECT tunnel is reported through the
authenticate signal of the original message, whose status is still
SOUP_STATUS_NONE at that point. Checking the message status for 407
therefore never matches, proxy-id/proxy-pw are never used, and HTTPS
through an authenticating proxy fails. Use soup_auth_is_for_proxy()
instead, which reflects where the challenge came from.
---
 ext/soup/gstsouphttpsrc.c | 13 ++++++++-----
 ext/soup/gstsouploader.c  | 13 +++++++++++++
 ext/soup/gstsouploader.h  |  3 +++
 3 files changed, 24 insertions(+), 5 deletions(-)

diff --git a/subprojects/gst-plugins-good/ext/soup/gstsouphttpsrc.c b/subprojects/gst-plugins-good/ext/soup/gstsouphttpsrc.c
index 91f9d11..42d8982 100644
--- a/subprojects/gst-plugins-good/ext/soup/gstsouphttpsrc.c
+++ b/subprojects/gst-plugins-good/ext/soup/gstsouphttpsrc.c
@@ -1405,14 +1405,17 @@ gst_soup_http_src_authenticate_cb (SoupMessage * msg, SoupAuth * auth,
   if (!retrying) {
     /* First time authentication only, if we fail and are called again with
      * retry true fall through */
-    if (status_code == SOUP_STATUS_UNAUTHORIZED) {
-      if (src->user_id && src->user_pw) {
-        _soup_auth_authenticate (auth, src->user_id, src->user_pw);
-      }
-    } else if (status_code == SOUP_STATUS_PROXY_AUTHENTICATION_REQUIRED) {
+    /* libsoup 3 reports a 407 from a CONNECT tunnel through the original
+     * message, which has no status yet, so ask the auth whether it is for
+     * the proxy. */
+    if (_soup_auth_is_for_proxy (auth)) {
       if (src->proxy_id && src->proxy_pw) {
         _soup_auth_authenticate (auth, src->proxy_id, src->proxy_pw);
       }
+    } else if (status_code == SOUP_STATUS_UNAUTHORIZED) {
+      if (src->user_id && src->user_pw) {
+        _soup_auth_authenticate (auth, src->user_id, src->user_pw);
+      }
     }
   }
 
diff --git a/subprojects/gst-plugins-good/ext/soup/gstsouploader.c b/subprojects/gst-plugins-good/ext/soup/gstsouploader.c
index 41f3aa3..8898401 100644
--- a/subprojects/gst-plugins-good/ext/soup/gstsouploader.c
+++ b/subprojects/gst-plugins-good/ext/soup/gstsouploader.c
@@ -128,6 +128,7 @@ typedef struct _GstSoupVTable
 
   void (*_soup_auth_authenticate) (SoupAuth * auth, const char *username,
     const char *password);
+  gboolean (*_soup_auth_is_for_proxy) (SoupAuth * auth);
   const char *(*_soup_message_get_method_3) (SoupMessage * msg);
   GInputStream *(*_soup_session_send_async_2) (SoupSession * session, SoupMessage * msg,
     GCancellable * cancellable, GAsyncReadyCallback callback, gpointer user_data);
@@ -222,6 +223,7 @@ gst_soup_load_library (void)
       }
 
       LOAD_SYMBOL (soup_auth_authenticate);
+      LOAD_SYMBOL (soup_auth_is_for_proxy);
       LOAD_SYMBOL (soup_content_decoder_get_type);
       LOAD_SYMBOL (soup_cookie_jar_get_type);
       LOAD_SYMBOL (soup_get_major_version);
@@ -807,6 +809,17 @@ _soup_auth_authenticate (SoupAuth * auth, const char *username,
 #endif
 }
 
+gboolean
+_soup_auth_is_for_proxy (SoupAuth * auth)
+{
+#ifdef LINK_SOUP
+  return soup_auth_is_for_proxy (auth);
+#else
+  g_assert (gst_soup_vtable._soup_auth_is_for_proxy != NULL);
+  return gst_soup_vtable._soup_auth_is_for_proxy (auth);
+#endif
+}
+
 const char *
 _soup_message_get_method (SoupMessage * msg)
 {
diff --git a/subprojects/gst-plugins-good/ext/soup/gstsouploader.h b/subprojects/gst-plugins-good/ext/soup/gstsouploader.h
index ccc7445..8138589 100644
--- a/subprojects/gst-plugins-good/ext/soup/gstsouploader.h
+++ b/subprojects/gst-plugins-good/ext/soup/gstsouploader.h
@@ -62,6 +62,7 @@
 #define _soup_message_headers_get_content_range _ad2_soup_message_headers_get_content_range
 #define _soup_message_headers_set_range _ad2_soup_message_headers_set_range
 #define _soup_auth_authenticate _ad2_soup_auth_authenticate
+#define _soup_auth_is_for_proxy _ad2_soup_auth_is_for_proxy
 #define _soup_message_get_method _ad2_soup_message_get_method
 #define _soup_session_send_async _ad2_soup_session_send_async
 #define _soup_session_send_finish _ad2_soup_session_send_finish
@@ -153,6 +154,8 @@ void _soup_message_headers_set_range (SoupMessageHeaders *hdrs, goffset start, g
 void _soup_auth_authenticate (SoupAuth *auth, const char *username,
                               const char *password);
 
+gboolean _soup_auth_is_for_proxy (SoupAuth *auth);
+
 const char *_soup_message_get_method (SoupMessage *msg);
 
 void _soup_session_send_async (SoupSession *session,
